### PR TITLE
Extend fluent-bit dashboard with custom metrics

### PR DESCRIPTION
**Pull request by Kristian-ZH**
**How to categorize this PR?**

/area logging
/kind enhancement
/priority normal

**What this PR does / why we need it**:
Upgrade output plugin version to `v0.29.0`
Include in the fluent-bit dashboard new metrics from the custom plugin.
The new fluent-bit image also includes a fix for the OutOfOrder Loki behaviour

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
@wyb1 @vlvasilev 

**Release note**:

```improvement operator
Sort logs to fix out of order issue
```
```improvement operator
The output plugin exposes custom metrics
```
```improvement operator
Modified fluent-bit dashboard to include the new metrics
```

### CODE REVIEW

**Review by vlvasilev — DISMISSED** [review]
/lgtm

File: charts/seed-bootstrap/charts/fluent-bit/templates/fluent-bit-daemonset.yaml [hunk 1]
```diff
@@ -56,6 +56,9 @@ spec:
         - name: metrics
           containerPort: 2020
           protocol: TCP
+        - name: metrics-plugin
+          containerPort: 2021
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/fluent-bit/templates/fluent-bit-daemonset.yaml]:
```suggestion
          containerPort: {{ .Values.ports.outputPluginMetrics }}
```

File: charts/seed-bootstrap/charts/fluent-bit/templates/fluent-bit-daemonset.yaml [hunk 2]
```diff
@@ -56,6 +56,9 @@ spec:
         - name: metrics
           containerPort: 2020
```
Comment (rfranzke) [inline charts/seed-bootstrap/charts/fluent-bit/templates/fluent-bit-daemonset.yaml]:
```suggestion
          containerPort: {{ .Values.ports.metrics }}
```

**Comment by Kristian-ZH** [comment]
Will convert this PR to draft because I would like to change a little bit the dashboard

**Comment by Kristian-ZH** [comment]
ready

**Review by rfranzke — DISMISSED** [review]
/lgtm
/needs second-opinion
/invite @wyb1 @istvanballok @vpnachev

**Review by vpnachev — DISMISSED** [review]
/lgtm

**Comment by vlvasilev** [comment]
There is bug with fluent-bit-to-loki 0.30 version where some times we close a wait group two times leading to panic.
Please do not merge with 0.30 version

File: charts/images.yaml [hunk 3]
```diff
@@ -204,7 +204,7 @@ images:
 - name: fluent-bit-plugin-installer
   sourceRepository: github.com/gardener/logging
   repository: eu.gcr.io/gardener-project/gardener/fluent-bit-to-loki
-  tag: "v0.28.0"
+  tag: "v0.30.0"
```
Comment (vlvasilev) [inline charts/images.yaml]: Bugged version of fluent-bit-to-loki plugin.

**Review by vlvasilev — CHANGES_REQUESTED** [review]

File: charts/images.yaml (hunk 3 above)
Comment (vpnachev) [reply charts/images.yaml]: /hold

**Review by wyb1 — DISMISSED** [review]
/lgtm

just out of curiosity when would these tables be different?

image

**Comment by Kristian-ZH** [comment]
> /lgtm
> 
> just out of curiosity when would these tables be different?
> 
> image

Wrong configurations for the Output plugin for example. Between the two metrics there are logic for removing unnecessary keys, getting client for the correct Loki  and  forming the log line.

If there is difference between the two tables, it indicates an issue in the output plugin so we will create also alerts for these metrics

**Comment by Kristian-ZH** [comment]
After that the log is processed to the Promtail client which sends it to the Loki. For this reason we monitor the Promtail dropped logs also

**Comment by rfranzke** [comment]
What's the status of this PR, and in particular, can you share details about https://github.com/gardener/gardener/pull/3188#discussion_r529608425 and the plan forward?

**Comment by Kristian-ZH** [comment]
> What's the status of this PR, and in particular, can you share details about [#3188 (comment)](https://github.com/gardener/gardener/pull/3188#discussion_r529608425) and the plan forward?



> What's the status of this PR, and in particular, can you share details about [#3188 (comment)](https://github.com/gardener/gardener/pull/3188#discussion_r529608425) and the plan forward?

We are waiting to release a new version of the logging where this bug is fixed. Now the release job is triggered.
The bug is that we were calling two sequential  times WaitGroup.Done().

**Comment by rfranzke** [comment]
Thank you @Kristian-ZH 👍

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Review by vpnachev — DISMISSED** [review]
/lgtm

**Review by vlvasilev — DISMISSED** [review]
/lgtm

**Comment by vpnachev** [comment]
@wyb1 @istvanballok could you take a look?

File: charts/seed-bootstrap/templates/aggregate-prometheus/config.yaml [hunk 4]
```diff
@@ -132,6 +132,28 @@ data:
...
 
+    - job_name: fluent-bit-output-plugin
+      honor_labels: false
+      kubernetes_sd_configs:
+      - role: endpoints
+        namespaces:
+          names: [ garden ]
+      relabel_configs:
+      - target_label: __metrics_path__
+        replacement: /metrics
```
Comment (istvanballok) [inline charts/seed-bootstrap/templates/aggregate-prometheus/config.yaml]: https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config
`/metrics` is the default metrics path, so you might be able to omit these 2 lines

File: charts/seed-bootstrap/values.yaml [hunk 5]
```diff
@@ -111,6 +111,15 @@ allowedMetrics:
   - fluentbit_filter_add_records_total
   - fluentbit_filter_drop_records_total
 
+  fluentbitOutputPlugin:
+  - promtail_dropped_entries_total
+  - output_plugin_errors
+  - output_plugin_no_metadata
+  - output_plugin_incoming_logs
+  - output_plugin_past_send_requests
+  - output_plugin_dropt_logs
```
Comment (istvanballok) [inline charts/seed-bootstrap/values.yaml]: I think it might be a bit more involved to fix this typo: `dropt`, as it is coming from the fluent-bit-to-loki plugin,
https://github.com/gardener/logging/blob/99b64c43b0a2e84e833b74d318cad8292431ad34/pkg/metrics/metrics.go#L54
but better now then later :) [drop -> dropped](https://www.oxfordlearnersdictionaries.com/definition/english/drop_1)

Comment (Kristian-ZH) [reply charts/seed-bootstrap/values.yaml]: Yes, I know already about this typo, but as you mentioned, it is not coming from this PR.

Comment (Kristian-ZH) [reply charts/seed-bootstrap/values.yaml]: So lets hold this PR until I release the logging repo with the renaming fix

**Comment by istvanballok** [comment]
Metric names are strings that are typically used (=hard coded) in diverse locations (recording rules, alerting rules, promql queries in Grafana dashboards), with no "compiler" support to assist with renames - so we can only dream about the comfortable situation when golang developers rename a variable with their IDE, and it just works. Furthermore, the motivation for scraping metrics is to store them for a few weeks at least, to show how the values change over time, which makes renaming metrics even more challenging (e.g. the promql query would have to support both names during the transition period, as long as the old metric names are still kept around according to the retention policy - otherwise, the old metrics would be lost for the visualization).
Long story short, after chatting with @Kristian-ZH, we concluded that it might be worthwhile to fix the typos and adhere to the [naming conventions](https://prometheus.io/docs/practices/naming/#metric-and-label-naming) (e.g. _total suffix for counters) for the very first release of the custom metrics feature.

**Review by vpnachev — APPROVED** [review]
/lgtm
/invite @istvanballok

File: charts/seed-bootstrap/values.yaml [hunk 6]
```diff
@@ -111,6 +111,14 @@ allowedMetrics:
   - fluentbit_filter_add_records_total
   - fluentbit_filter_drop_records_total
 
+  fluentbitOutputPlugin:
+  - promtail_dropped_entries_total
+  - fluentbit_loki_gardener_errors_total
+  - fluentbit_loki_gardener_logs_without_metadata_total
+  - fluentbit_loki_gardener_incoming_logs_total
+  - fluentbit_loki_gardener_forwarded_logs_total
+  - fluentbit_loki_gardener_dropped_logs_total
```
Comment (istvanballok) [inline charts/seed-bootstrap/values.yaml:120]: 👍

**Review by istvanballok — APPROVED** [review]
/lgtm
